packages/unified-selection/src/unified-selection/SelectionScope.ts

```diff
@@ -0,0 +1,279 @@
+                      UNION ALL
+                      SELECT pe.ECInstanceId, pe.ECClassId
+                      FROM Element2dTopAssemblyElements AS metae
+                        JOIN BisCore.Element e ON e.ECInstanceId = metae.ECInstanceId
+                        JOIN BisCore.Element pe ON pe.ECInstanceId = e.Parent.Id
+                    ),
+                    Element2dTopAssemblyRelatedFunctionalElements (ECInstanceId, ClassName) AS (
+                      SELECT COALESCE(r.TargetECInstanceId, metae.ECInstanceId), ec_classname(COALESCE(r.TargetECClassId, metae.ECClassId), 's.c')
+                      FROM Element2dTopAssemblyElements metae
+                        LEFT JOIN Functional.PhysicalElementFulfillsFunction r ON r.SourceECInstanceId = metae.ECInstanceId
+                    ),
+                    Element3dTopAssemblyElements (ECInstanceId, ECClassId) AS (
+                      SELECT ECInstanceId, ECClassId FROM BisCore.GeometricElement3d WHERE ${formIdBindings("ECInstanceId", ids, bindings)}
+                      UNION ALL
+                      SELECT pe.ECInstanceId, pe.ECClassId
+                      FROM Element3dTopAssemblyElements AS etae
+                        JOIN BisCore.Element e ON e.ECInstanceId = etae.ECInstanceId
+                        JOIN BisCore.Element pe ON pe.ECInstanceId = e.Parent.Id
+                    ),
+                    Element3dTopAssemblyRelatedFunctionalElements (ECInstanceId, ClassName) AS (
+                      SELECT COALESCE(r.TargetECInstanceId, etae.ECInstanceId), ec_classname(COALESCE(r.TargetECClassId, etae.ECClassId), 's.c')
+                      FROM Element3dTopAssemblyElements etae
+                        JOIN BisCore.Element e ON e.ECInstanceId = etae.ECInstanceId
+                        LEFT JOIN Functional.PhysicalElementFulfillsFunction r ON r.SourceECInstanceId = etae.ECInstanceId
+                      WHERE e.Parent.Id IS NULL
+                    )
+                    SELECT * FROM Element2dTopAssemblyRelatedFunctionalElements
+                    UNION
+                    SELECT * FROM Element3dTopAssemblyRelatedFunctionalElements;`;
+
+    yield* executeQuery(queryExecutor, query, bindings);
+  }
+
+  function getNearestFunctionalElementQuery(ids: string[], bindings: ECSqlBinding[]): string {
+    return `Element2dNearestFunctionalElement (OriginalECInstanceId, OriginalECClassId, ECInstanceId, ECClassId, ParentId) AS (
+              SELECT ECInstanceId, ECClassId, ECInstanceId, ECClassId, Parent.Id FROM BisCore.GeometricElement2d WHERE ${formIdBindings("ECInstanceId", ids, bindings)}
+              UNION ALL
+              SELECT e.OriginalECInstanceId, e.OriginalECClassId, COALESCE(r.TargetECInstanceId, ge.ECInstanceId), COALESCE(r.TargetECClassId, ge.ECClassId), ge.Parent.Id
+              FROM Element2dNearestFunctionalElement e
+                LEFT JOIN BisCore.Element ge ON ge.ECInstanceId = e.ParentId
```

Comment:
Please try to come up with aliases that help read the query. Generally, `ge` is used for geometric or graphical elements, but you're using it for a parent element. I'd rather see `p`, `pe` or even `parent` instead of `ge`.

Reply:
Made the aliases more relevant.